### PR TITLE
Implement fallback image - issue #493

### PR DESCRIPTION
**Pull request by JStein92**
## This PR is a:

- [x] New feature
- [x] Enhancement/Optimization
- [x] Refactor
- [ ] Bugfix
- [ ] Test for existing code
- [ ] Documentation

images-online

## Summary

When this pull request is merged, it will allow images to render a placeholder and 'error/offline' icon through a new `Image` component.

- Add a generic `Image` component which can render a custom `placeholder`

- Replace the `img` tag in `item.js` with our generic `Image` component

- Handle image errors through the new `Image` component by rendering the icon shown in #493 on top of the placeholder

- Transitioning from offline to online will trigger images using the `Image` component to re-render. 

## Additional information

-  The `error` icon is being stored locally - we're storing it with Workbox-webpack it so it can always display offline

- We decided to make the `placeholder` a prop passed to the `Image` component so that custom placeholders could be rendered - ex. a circle placeholder image for circular category images

- `iconHeight` is used to set the `error` icon's height, since it has set pixel heights depending on the image size - see @soumya-ashok's comments of #493

### CODE REVIEW

**Comment by supernova-at** [comment]
Hey @JStein92 , just catching up. Is this good to review? Last comment was on January 3rd - I know a lot has changed since then =)

**Comment by JStein92** [comment]
@supernova-at I'll get this up to date and ping you when it's ready to look at again :)

**Comment by pcvonz** [comment]
Removed a check for an empty item in `gallery/item.js`, seems like it's not necessary with the way the loader works now(?)

**Comment by supernova-at** [comment]
> Removed a check for an empty item in gallery/item.js, seems like it's not necessary with the way the loader works now(?)

I think if the `data.category.products.items` in [categoryContent.js](https://github.com/magento-research/pwa-studio/blob/develop/packages/venia-concept/src/RootComponents/Category/categoryContent.js#L26) is an empty array, we still may hit the codepath in `Gallery/item.js` where `props.item` is `undefined`.

For that reason I added back in some null checking to render placeholders when `item` is not provided.

I tested this by changing L27 in `Gallery/gallery.js` to simulate that scenario:

```js
// const items = hasData ? data : emptyData;
const items = emptyData;
```

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento-research/pwa-studio?pullRequest=672) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dev.magento.com/magento-research/pwa-studio?pullRequest=672) before we can accept your contribution.<br/>

**Comment by supernova-at** [comment]
@tjwiebell the main feature here is

> Transitioning from offline to online will trigger images using the Image component to re-render.

To Repro:
1. Hit main / CMS page
1. Click "Tops" category
1. Go offline
1. Click to next page
1. Observe broken images. With this PR they look nicer (hanger instead of browser broken image).
1. Go back online. With this PR the images re-load themselves.

**Current Behavior**
current

**With this PR**
new

File: packages/venia-concept/src/sw.js [hunk 1]
```diff
@@ -28,7 +28,7 @@ workbox.routing.registerRoute(
 );
 
 workbox.routing.registerRoute(
-    /\.(?:png|gif|jpg|jpeg|svg)$/,
+    /\/icons\/\.(?:png|gif|jpg|jpeg)$/,
```
[the remark this answers is not in the document]
Comment (sirugh) [reply packages/venia-concept/src/sw.js]: @JStein92 did you see this question?

**Comment by sirugh** [comment]
Looks like being offline actually breaks the app. If you check out `develop` and do the same steps you posted above you get a "Data fetch error" when you click to the next page:

[Image from Gyazo](https://gyazo.com/b6ae7f41d6623a64b9f4cb16c34f2ade)

**Comment by JStein92** [comment]
Going to let @sirugh take over this PR, since I'm not currently able to work on Venia.

**Comment by sirugh** [comment]
After reviewing this PR I think the best way to proceed will actually be to close the PR for several reasons:

1) Currently Venia doesn't seem to even display subsequent category pages after going "offline". To repro, open a category, go offline, and try to navigate to the next page in the category.
2) The existing image placeholder styling is sufficient and a good UX. When you open a category for the first time you will see a placeholder image along with placeholder boxes where the item name and price would exist. These are replaced by the item as soon as the item loads. In this PR we replaced those placeholders entirely with only an image placeholder.
3) I believe there are several paths forward in terms of handling assets which are not loaded. In terms of images we may want an `Image` component that takes a `placeholder` prop along with styling info so that we can render a fallback image should we get a 404.
4) Tying the `Image` component to the app's online state seems odd. If the image is already fetched we don't want to render a placeholder image even if the app is offline.

I'll keep the original issue (#493) open but we will need to revisit it to see if it is still a valid scenario.

**Pull request closed without merging**